pm/queue.go

```diff
@@ -140,12 +126,17 @@ func (q *ticketQueue) startQueueLoop() {
 	}
 }
 
-// removeHead removes the head of the queue
-func (q *ticketQueue) removeHead() {
-	// Lock conditional variable while removing from the queue
-	q.cond.L.Lock()
-	q.queue[0] = nil
-	q.queue = q.queue[1:]
-	atomic.AddInt32(&q.queueLen, -1)
-	q.cond.L.Unlock()
+// pop returns the head of the queue
+func (q *ticketQueue) pop() (*SignedTicket, error) {
+	ticket, err := q.store.SelectEarliestWinningTicket(q.sender)
+	if err != nil {
+		return nil, err
+	}
+	if ticket == nil {
+		return nil, nil
+	}
+	if err := q.store.StoreWinningTicket(ticket, true); err != nil {
```

Comment:
Instead of overloading the usage of `StoreWinningTicket()` here, could we:

- Update `RemoveWinningTIcket()` to mark a ticket as submitted
- Return the earliest winning ticket in `pop()`
- Wait for the ticket to be submitted before selecting the next earliest winning ticket in the queue loop
- If the ticket was submitted without an error, call `RemoveWinningTicket()` for the ticket
- At this point, select the next earliest winning ticket

This would keep the sole responsibility for `StoreWinningTicket()` as writing a new ticket to the store, changes the "delete" operation in `RemoveWinningTicket()` to just update an existing ticket as submitted and avoids updating the store when a ticket needs to be retried.

The implementation might look something like:


```golang
type redeemSignedTicket struct {
    *SignedTicket
    errCh chan error
}

func (q *ticketQueue) startQueueLoop() {
    ...
    if nextTicket.ParamsExpirationBlock.Cmp(latestBlock) <= 0 {
        errCh := make(chan error, 1)
        select {
            case q.redeemable <- &redeemSignedTicket{SignedTicket: nextTicket, errCh: errCh}
                // Make sure that the consumer of q.redeemable uses errCh to pass an error value
                // back (nil indicates success)
                if err := <-errCh; err != nil {
                    // Remove ticket
                } else {
                    // Handle error
                }
        }
    }
}
```

Reply:
I think there's two separate discussions here 

- Using UPSERT vs Using two queries 
- Using an error channel vs having the consumer be responsible for re-adding tickets upon error

RE UPSERT: 
1. It's an UPSERT SQL query , I don't see how that's overloading. Having to create an additional query for this would be overloading the API. 

2. The function will have a totally different functionality than it's description (not insurmountable, but see 1)

3. The RemoveWinningTicket function should be reserved for DB pruning and actually deleting a winning ticket. For the same reason the method has not been removed despite being unused

4. The sole responsibility characteristic of `StoreWinningTicket` as an UPSERT query does not change. It stores winning tickets and their status accordingly.


RE ErrCh 
The behaviour described is not thread-safe, _I think_,  since tickets are not marked submitted at the moment they are sent into the channel. While this is not an issue right now, it is if we were to allow such behaviour in the future. 

I do see the benefit because in case the node shuts down while the current ticket is still in flight it will not be marked submitted and automatically retried when the node comes back online. _But_ this seems like a very rare edge case, especially when using the redemption service (basically when the user shuts down the node or if there's an unexpected panic).

So I'm wondering if this edge case warrants completely changing the behaviour of the existing implementation.

Reply:
On UPSERT, it's also a matter of being clear about intention. UPSERTs are useful if you aren't sure if the value exists in the table. In this case, we know that a value will will always exist. (Maybe it error out otherwise, but I'm indifferent.) The `RemoveWinningTicket` function can be renamed to `MarkTicketSubmitted` or something and the query updated accordingly. That makes the effect of "marking a ticket as submitted" more straightforward for the reader.

Another way to consider it is, there is no overlap in the fields that are actually updated between the initial insert and the update. The `submitted` field can be omitted from the initial insert, and updated from a separate query. This preserves the clarity of intent.

> So I'm wondering if this edge case warrants completely changing the behaviour of the existing implementation.

I'm not totally clear on when a ticket is considered "submitted" versus "in-flight" but will #654 mean accommodating this particular edge case? If so, it might be good to make the suggested change anyway, as a first step to avoid re-working code later on once #654 is implemented.

Reply:
In the current implementation we mark a ticket as submitted as soon as it's sent to the consumer. 

In [this comment](https://github.com/livepeer/go-livepeer/pull/1493#discussion_r434041781) a ticket would only be marked submitted once the consumer returns an error (or `nil`), so with in-flight I mean the period between sending it to the consumer and the moment we mark it as submitted. This would also put the responsibility of re-adding tickets to the queue for errors back with the queue, while the consumer already does some pre-flight checks so we'd have to start checking the returned error as well.

Reply:
> On UPSERT, it's also a matter of being clear about intention. UPSERTs are useful if you aren't sure if the value exists in the table.

I was looking at the implementation details. An UPSERT would actually be more favorable if we don't change the current implementation of the queueloop to also receive errors on a channel because tickets are added back to the queue by the consumer on error. If we don't use an UPSERT `Queue.Add` would have to check if the ticket already exists. 

Thus my compromise proposition is 
- Don't change queue loop behaviour for now
- Use a `submittedAt` timestamp (set it to time.Now if we sent a ticket to the consumer over the channel, set it back to NULL if a ticket gets re-added to the queue but already exists)
- keep the UPSERT to avoid an extra check in `queue.Add`

Reply:
> Because of this "in-flight" status we wouldn't be able to run the inner loop concurrently if we'd ever choose to (e.g. if we receive new block events faster than we can run through the loop).

Note that the inner loop you describe (the one that iterates through all the tickets in the queue) already cannot progress faster than the block time if the queue is not empty. At the moment, the redeemable channel consumer redeems a single ticket at a time and redemption takes at least a single block. So, by the time the channel consumer is ready for the next ticket, a new block would have been received, but before the ticket loop can handle the new block the inner loop would need to finish.

>  If we don't use an UPSERT Queue.Add would have to check if the ticket already exists.

Wouldn't this be a non-issue if we only update the timestamp after the inner loop receives confirmation that the ticket was submitted (if the confirmation is sent after the `redeemWinningTicket()` method returns then this would mean the tx was also mined)? Then, we wouldn't have to worry about re-adding tickets to the queue. 

From my POV, it seems that waiting for confirmation in the inner loop is a small change that only has benefits i.e. avoiding re-adding logic and also addressing the edge case that you mentioned. Let me know if I'm missing anything though.

I also agree with Josh about the better clarity of intent with a separate method/query (maybe `UpdateWinningTicket()`)  for updating the timestamp instead of the `StoreWinningTicket()` method and an UPSERT query.

Reply:
> so if there's case where a queue receives multiple winning tickets per block then it's playing catch up until it doesn't receive winning tickets from that sender for a while.

I'm not sure if the pace of ticket redemption not matching the rate at which blocks are received is a problem - do you see any issues that may arise? Furthermore, one other option we have if we do believe that the pace of ticket redemption needs to be faster is to redeem a batch of tickets in a single tx (using all the tickets from the queue).

Reply:
> If we don't use an UPSERT Queue.Add would have to check if the ticket already exists.

Attempting to INSERT on an already-existing row should fail (primary key collision) so we can just return that error, rather than doing an explicit check?

Reply:
fixed in 726da54ac56c1724864ca99e57791dfc936148fc and 5783c60038ba7cf9e2ceb9226c622d5dbba299ab

I didn't decide to bother fixing up the `recipient` as the end consumer of the ticketqueue because that gets removed in the subsequent PR so I just made sure the unit tests pass for now and will fix up the remainder of the implementation in #1495